### PR TITLE
peepmatic: Add bnot operation

### PR DESCRIPTION
**Pull request by MaxGraey**
- Add `Bnot` operation for peepmatic.
- ~~Add `x ^ -1` -> `~x` rule for preopt.peepmatic~~

### CODE REVIEW

File: cranelift/filetests/filetests/peepmatic/simplify64.clif [hunk 1]
```diff
@@ -324,3 +324,17 @@ block0(v0: i32):
...
+    ; check: bnot v0
+    return v1
+}
+
+function %bxor_imm_64(i64) -> i64 {
+block0(v0: i64):
+    v1 = bxor_imm v0, -1
+    ; check: bnot v0
+    return v1
+}
```
Comment (bjorn3) [inline cranelift/filetests/filetests/peepmatic/simplify64.clif:340]: nit: missing trailing newline

**Comment by MaxGraey** [comment]
> I do think that Peepmatic support for bnot instructions is useful, so if you
would be interested in separating out just the changes required to support
bnot, then I would be happy to merge them.

What if I just add support same transform also for `simple_preopt.rs` for parity?

**Comment by fitzgen** [comment]
> What if I just add support same transform also for `simple_preopt.rs` for parity in this PR?

`simple_preopt.rs` optimizations need to be careful to carry their weight (which is part of what peepmatic is trying to improve upon) and I'm not sure what the exact calculus is that makes an optimization worth it or not. Perhaps @sunfishcode or @bnjbvr could comment on whether this would be a welcome addition to `simple_preopt.rs`.

If so, then yes let's add it  `preopt.peepmatic` as well!

**Comment by bnjbvr** [comment]
Thanks for working on this!

> simple_preopt.rs optimizations need to be careful to carry their weight (which is part of what peepmatic is trying to improve upon) and I'm not sure what the exact calculus is that makes an optimization worth it or not. Perhaps @sunfishcode or @bnjbvr could comment on whether this would be a welcome addition to simple_preopt.rs.

The simple preopt pass was designed to catch frequent patterns that lead to better code generation and that were not directly produced by the wasm producer. We expect most of these optimizations to have already happened in e.g. LLVM. But the IR translation makes it so that some new pattern-matching opportunities on the CLIR arise. Most of simple preopt phase was mostly acting as an helper for instruction selection (legalization) that happened later in the pipeline, by selecting the `_imm` variants so that legalization would use machine instructions with immediates; the "pre" in preopt refers to the fact that it runs before legalization. However, instruction selection happens in a totally different place now, with the new target backends, so this part is becoming less useful.

Integer transforms like could be valuable at the `speed` optimization level; there's some overhead for pattern-matching every single instruction, so we try to be careful to not put every optimization we can think of. Usually our line of thinking has been:

- is it reasonable that the CLIF producer already performed such an optimization? If yes, do no replicate it in Cranelift.
- data-driven approach: for large CLIF inputs (from large wasm programs), how many times is the optimization effective? if it's very frequent (there's no "official" threshold, so common sense is required here), then do it in Cranelift.

For instance, constant propagation doesn't quite happen in Cranelift, because we really expect the producer to have done it. For other producers which don't have advanced optimization passes like this, there's an external crate for `cranelift-preopt` that does exactly this.

So considering the answers to the two questions above, it might be worth putting it in simple_preopt, or implementing it in the `preopt` crate?

**Comment by MaxGraey** [comment]
@bnjbvr thanks for explanation!
Main goal of this PR is speedup code produced by webassembly. As you know wasm doesn't have `not` operation and emulate it via `x ^ -1`. Also perhaps it could be useful even for LLVM code built with [`-O0` for some reasons](https://godbolt.org/z/1z9r16).

**Comment by julian-seward1** [comment]
Also, for something as simple as `x ^ -1`, the new backend(s) that are under development, could easily enough match that in their instruction selectors and directly emit a `not` insn as appropriate.  There's no requirement for any more heavyweight machinery to handle this particular case.

**Comment by MaxGraey** [comment]
Hmm, could you explain which kind of rules are preferable for peepmatic DSL? It would really help me.

**Comment by MaxGraey** [comment]
Alright. I removed rule but keep `bnot` instruction code for peepmatic

**Pull request merged**